### DISCUSSION TITLE
How to add adapters and configurations to dap.core

### DISCUSSION THREAD

**Discussion by kyrillh** [LazyVim/LazyVim#894] [Q&A] [answered]
I have imported the shipped dap configuration by adding 
```
{ import = "lazyvim.plugins.extras.dap.core"},
``` 
to `lua/config/lazy.lua`. 
To add a debug adapter and configuration for C++ I added the following to the root `init.lua` file
```
  require("config.lazy")
  local dap = require("dap")
  dap.adapters.cppdbg = {
      id = "cppdbg",
      type = "executable",
      command = "path/to/OpenDebugAD7",
  }
  dap.configurations.cpp = {
      {
          name = "Launch file",
          type = "cppdbg",
          request = "launch",
          program = "/path/to/debugee",
          cwd = "${workspaceFolder}",
          stopAtEntry = true,
      },
  }
```
This works great, but does not conform to how plugin options are set for lazyVim. 
How can I set these tables in a configuration file for dap under `~/.config/nvim/lua/plugins`? 
Thanks!

**Comment by kyrillh** [comment]
Update: 
I have tried adding a `dap.lua` file in `~/.config/nvim/lua/config/` with 
```
local dap = require("dap")
  dap.adapters.cppdbg = {
      id = "cppdbg",
      type = "executable",
      command = "path/to/OpenDebugAD7",
  }
  dap.configurations.cpp = {
      {
          name = "Launch file",
          type = "cppdbg",
          request = "launch",
          program = "/path/to/debugee",
          cwd = "${workspaceFolder}",
          stopAtEntry = true,
      },
  }
```
but this does not update the `dap.adapters` and the `dap.configurations` tables. 
I am new to neovim and lua and am a bit confused by why this doesn't update the tables. 
I thought about adding a `dap.lua` file in `~/.config/nvim/lua/plugins` with something like
```
return{
  local dap = require("dap")
    dap.adapters.cppdbg = {
        id = "cppdbg",
        type = "executable",
        command = "path/to/OpenDebugAD7",
    }
    dap.configurations.cpp = {
        {
            name = "Launch file",
            type = "cppdbg",
            request = "launch",
            program = "/path/to/debugee",
            cwd = "${workspaceFolder}",
            stopAtEntry = true,
        },
    }
}
```
(obviously invalid syntax), but wasn't sure how to access the tables here. Any suggestions would be greatly appreciated. Thanks!

**Comment by JeffDess** [comment, marked as answer]
I was quite puzzled about adding custom adapters and configs too. I added `/plugins/nvim-dap.lua` : 

```lua
return {
  "mfussenegger/nvim-dap",
  opts = function()
    local dap = require("dap")
    dap.adapters.node2 = {
      type = "server",
      ...
    }
    for _, language in ipairs({ "typescript", "javascript" }) do
      if not dap.configurations[language] then
        dap.configurations[language] = {
          {
            type = "node2", 
            name = "My Custom Config"
             ...
          },
        }
      end
    end
  end,
}
```
Running this without dap.core in `config/lazy.lua`, I would see "My Custom Config" when running `require("dap").continue()`. But when I imported dap.core, it wasn't showing up in the options, it was replaced by what I have installed with mason. 

The trick is to organize the plugin file like this:

```lua
return {
  "mfussenegger/nvim-dap",
  dependencies = {
    {
      "jay-babu/mason-nvim-dap.nvim",
      event = "BufReadPre", -- <-- this
    },
  },
  opts = function()
    ...
  end
}
```

This way, everything will show up in the `dap.continue()` menu, both mason configs and what you have defined in your plugin file.

**Reply by kyrillh** [reply]
Excellent! Thanks for this. Interestingly your first solution worked for me too i.e. my configurations overwrite those imported by mason. However, with the `mason-nvim-dap` plugin I can now choose between my configs and the default ones which is nice.
